### ISSUE TITLE
[FEATURE] New command to copy files from/to VMs

### ISSUE THREAD

**Issue by mwieczorek** [Azure/azure-cli#5275]
Is it possible to add feature to azure cli to copy files from/to VMs?
Something like:
```
az vm upload --resource-group rg --name vmname --from <localpath> --to <reomtepath> 
```

**Comment by yugangw-msft** [comment]
If your VM has a public ip address, you have lots of good tools to upload a file. `az vm open-port` can also be used if the tool you prefer requires a particular port to be opened.

**Comment by mwieczorek** [comment]
Thanks @yugangw-msft 
Sure I can upload files using SCP, but using az cli would make some automation tasks simpler, especially for VMs without public IP.

**Comment by squillace** [comment]
@yugangw-msft I'd love it if we could consider this. It's a great feature, because for automation you don't want to use all those other tools AND you handle the non-public cases.

**Comment by yugangw-msft** [comment]
@squillace, let me see how much i can do

**Comment by squillace** [comment]
Thanks so much, of course. Lemme know. To flesh out the scenario a bit, the more basic functionality that is **roughly speaking always required** in real world scenarios can be shortened in any way, the more useful the CLI becomes and the happier the customer is with it. This one would replace getting the ssh endpoint, logging in and/or retrieving the credentials from KeyVault (where in the real world they should likely exist), and THEN using scp -- and that assumes an external ssh endpoint (which in the real world many people will simply not want). 

yet the command is simple, and clean, and one of the critical tasks you will perform with a vm. In short, it's a real saver no matter whether you're automated or not.

**Comment by yugangw-msft** [comment]
Thanks for all the feedback. I have following ideas for review:
   1. This command can implement most of SCP's functionalities, so I would just name as `az vm scp`
   2. The command would expose following flags
       ```bash
        a. --ssh-private-key: default to `~/.ssh/id_rsa` since `vm/vmss create` also references it. It can take a file path, or keyvault secret id.
        b: --action <upload/download>: default to upload
        c. --local <file1> <file2> <folder1> (if download, then only accepts a single file path or directory, as many to many would be hard to get it right through command line)
        d. --remote <file1> <file2> <folder>  (if upload, then only take a single file path or directory)
        e: --recursive (when directory involved)
        f: --port (in case the port is different) 
        ```
    3. Examples:
       ```bash
        az vm -g rg1 -n vm1 --local ~/foo1.txt --remote ~/dest-dir
        az vm -g rg1 -n vm1 --action download  --local ~ --remote ~/dest-dir/bar.exe
        az vm -g rg1 -n vm1 --local ~/myweb/src --remote ~/websites --recursive
        az vm -g rg1 -n vm1 --local ~/myweb/src --remote ~/websites --recursive --ssh-private-key https://mmadeclitestkv.vault.azure.net/secrets/ssh-privates-key
       ```
Let me know any comments/suggestions; otherwise i will start to get things going

**Comment by squillace** [comment]
@khenidak that looks good to be at first blush. Whatchya think?

**Comment by khenidak** [comment]
are we wrapping ssh? i think user expect this to work in private (non exposed to internet) VMs

**Comment by khenidak** [comment]
@kkmsft  FYI..

**Comment by squillace** [comment]
Also, sure seems like a dupe of #5622 to me....

**Comment by squillace** [comment]
two birds one stone!

**Comment by khenidak** [comment]
While i see ssh of value.. short cuts for connection and so on. The above request is not `scp` wrapper. User should be able to copy files from/to an Azure VM using only the cli without any configuration on the VM other than `walinux` agent. The entire process should be done via the fabric. i.e if i have `contributor` access to the VM resources (i can reset the password, right?) then i should be able to copy from/to vm

**Comment by squillace** [comment]
agree ^^

**Comment by squillace** [comment]
I'm just wondering, rather, that #5622 _scenario_ would be largely solved by this. I don't want scp embedded here, either. Not the goal. Down/uploading easily through the fabric is the goal.

**Comment by yugangw-msft** [comment]
The implementation will not be the wrapper of SSH/SCP, and no dependencies on any external command. I chose the naming of `scp` just for better discoveries

**Comment by squillace** [comment]
That makes sense, but I'm concerned that if you do that and don't use `copy`, you'll get two things:
1. hey! This doesn't do everything scp does!
2. Hey! I just wanna copy files easily, what's the command for that? (newbies to the environment)

My guess is that copy-files or copy is best. @khenidak @kkmsft what do you think would be most obvious here? Do you think my concerns are valid?

**Comment by yugangw-msft** [comment]
For other questions, suggestions...
For the key/password, it can either come from local `~/.ssh`, command arguments,  keyvault-key, or you can run "az vm user" to reset them and pass the new one to the command.
For private vm instances inside VMSS, by default, CLI enables inbound nat rule to enable SSH protocol, so the same command should work, but we will need to test it 
For #5622, I am not convinced we need to support that much like ssh. External tools, to me, is sufficiently easy to use. I am fine to do `scp` functionality because it can help scripting. 
A question here though, what does `through the fabric` mean here, particularly the `fabric` ?

**Comment by squillace** [comment]
For #5622, I think this MOSTLY solves the user's need.
by "through the fabric" here I mean what @khenidak refers to above in https://github.com/Azure/azure-cli/issues/5275#issuecomment-412147656: `if I have contributor access to the VM resources... I should be able to copy from/to vm." authentication/auth is run through ARM, is the way I would say it. Make sense? I'm quite sure you're doing this, but just to be clear....

**Comment by squillace** [comment]
@yugangw-msft where are we on this? As it's not an scp wrapper, and we don't want that, and I really don't think we should call it "scp", as that will drive users mad with irritation when it doesn't DO scp, but if it wraps scp, then it won't do what we're asking. 

I note that I didn't respond to your comment above, `external tools, to me, is sufficiently easy to use. I am fine to do scp functionality because it can help scripting.` external tools are bad in this case. this ability, to use az to copy files through the aad auth mechanism onto the vm without any other tool or switch is what we're looking for. It's not about anything specific, either. Happy to write the formal spec if necessary.

**Comment by yugangw-msft** [comment]
I am pretty flexible on the naming thing :), as long as it is reasonable.
But we need to sort out how to implement this first. The most important scenario is to find the solution to exchange files with VMs w/o public ip. W/o SSH keys/passwords involved, one of the few options left, is the "run-command" API (a wrapper around the custom script extension). To get the file out there are 2 mechanisms I can think of"
    1.  @kkmsft is using `cat` and `dt` to exchange the file content through stdout. This is a bit tricky and is  also hitting the API limitation that only 4K stdout can be returned from the API, so you have to split up a big file manually.
    2. Use a common file share like storage accounts, as the bridge. CLI can generate a sas toke from the account and let the script in the VM upload to there, and then CLI will pull it down to local.

Other means you can think of? I am counting on you :)

**Comment by squillace** [comment]
I'm up to the challenge! Gimme a minute. :-)

**Comment by hofnarwillie** [comment]
@squillace / @yugangw-msft - any update on this?

**Comment by yonzhan** [comment]
We need compute service support to copy large file content from client machine to VM.

**Comment by Jeeveshk** [comment]
Has this been implemented? If so it would have been of great use. 
As we had plans of using the "az vm scp" command within our existing GitLab CI/CD.

**Comment by adiabuk** [comment]
Any updates on this?  Seems like an important feature to have

**Comment by ealkalai** [comment]
It's a bit quiet here. Any updates on this?
As a workaround I'm base64 encoding the file, then i run a command to decode and save to a file. Do **not** use this, method if the content of the file is confidential, since it will appear in logs

sample powershell scripts on the host
```
$fileInBase64=[System.Convert]::ToBase64String([System.Text.Encoding]::ASCII.GetBytes($fileContent))
$parms = @{
     'ResourceGroupName' = $resourceGroupName;
     'Name'              = $vmName ;
     'CommandId'         = 'RunShellScript' ;
     'ScriptPath'        = "decodeToFile.sh";
     'Parameter' = @{"arg1" = "$($fileInBase64)" ; "arg2"= "$($targetFileName)" }
}
$result = Invoke-AzVMRunCommand @parms
```

content of decodeToFile.sh:
```
#!/usr/bin/env bash

base64str=$1
location=$2

[ -z $base64str ] && echo "No base64 provided" && exit 1
[ -z $location ] && echo "No location provided" && exit 1

echo $base64str | base64 -d > $location
```

**Comment by Plaquevent** [comment]
After reading @yugangw-msft answer, I implemented this for my needs (send a setup shell script to my VMs after I create it using ARM Templates and calling them from Templates in a VM extension) : 

sas_token_expiry=`date -u -d "10 minutes" '+%Y-%m-%dT%H:%MZ'`
az storage account create --name "$sa_name" --resource-group "$rg_name" --location "$location" --sku "Standard_LRS"
sa_cs=`az storage account show-connection-string -g "$rg_name" -n "$sa_name"`
az storage share create --name "$fs_name" --account-name "$sa_name" --connection-string "$sa_cs"
az storage file upload -s "$fs_name" --source "$setup_file" --connection-string "$sa_cs"
sas_token=`az storage account generate-sas --expiry "$sas_token_expiry" --permissions "r" --resource-types "sco" --services "bqtf" --account-name "$sa_name" --ip "$vm_ip" --connection-string "$sa_cs"`
sas_token=$(echo "$sas_token" | jq -r '.')
setup_file_download_link="https://$sa_name.file.core.windows.net/$fs_name/$(basename $setup_file)?$sas_token"

This is far from being perfect but it can help waiting for an official functionality :)

**Comment by Segaras** [comment]
Is this copy a file to my VM request forgotten?
It would be really helpful if there was a comman like "Invoke-AzVMRunCommand".
Maybe like: Invoke-AzVMCopyFile -ResourceGroupName 'rgname' -VMName 'vmname' -FileSourcePath 'Path/File' -FileDestinationPath 'Path/'
@yugangw-msft Do you know anything about if this will be once available?

**Comment by zhoxing-ms** [comment]
> Invoke-AzVMCopyFile -ResourceGroupName 'rgname' -VMName 'vmname' -FileSourcePath 'Path/File' -FileDestinationPath 'Path/'

@Segaras Hi, in fact, this is a PowerShell command. if you have requirements related to PowerShell, please submmit issue to https://github.com/Azure/azure-powershell/issues

Whatever, I can understand what you want to achieve, which requires the service team to support a REST to upload the file to VM from local.

**Comment by zhoxing-ms** [comment]
> We need compute service support to copy large file content from client machine to VM w/o public ip.

@fitzgeraldsteele Could you please take a look at this feature request?

**Comment by fitzgeraldsteele** [comment]
Adding @Chase ***@***.***> from Client Tools PM team, @Ankit ***@***.***> from VM PM team who may have more informed opinion on this one.



Jerry
Steele
Microsoft
senior program manager
425.421.2566


________________________________

**Comment by delian** [comment]
I often have setup with VM which are isolated from internet and this feature is very valuable

**Comment by KarsonAlford** [comment]
It makes me cringe to be doing it this way.  The PS1 file cannot have single quotes or $_. 

For instance, this is acceptable in the file:
`    $machine = $machines.Where({ $PSItem.Uri -eq $hostUri })
`
This is not:
`    $machine = $machines | Where-Object { $_.Uri -eq $hostUri }
`


```
- task: AzureCLI@2
      displayName: 'Copy VM Script'
      inputs:
        azureSubscription: '$(AzureResourceManagerSC)'
        scriptType: 'pscore'
        scriptLocation: 'inlineScript'
        inlineScript: |
          $lines = Get-Content -Path './Pipelines/Environments/VMSetup.ps1'

          $command = { 
            $lines = <VmSetup>
            $path = 'c:\temp\VmSetup.ps1'
            $lines | Out-File -FilePath $path
          }

          $commandScript = $command.ToString() -replace "<VmSetup>", ("@('" + ($lines -join "',  '") + "')")
          $commandScript = $commandScript.ToString() -replace """", """"""

          Write-Host $commandScript

          az vm run-command invoke --resource-group "$(ResourceGroupName)" --name "$(VMName)" --command-id RunPowerShellScript --scripts $commandScript

```

**Comment by QuintinWillison** [comment]
```sh
% az scp --help
'scp' is misspelled or not recognized by the system.
```

😞 (2024)

**Comment by alharald** [comment]
bump... +100 to add this feature, this is not only useful for private Virtual Machines in Azure but also for Azure Arc

**Comment by kamildzi** [comment]
Google made it properly. Microsoft messed it up. Well, in that case, SSH over Azure CLI tools is simply unusable. It's not complete.

**Comment by Ashutosh0x** [comment]
Hi @fitzgeraldsteele @qwordy @zhoxing-ms,

It's been a while (2018!), but the demand for this feature hasn't waned. As @squillace highlighted years ago, the core scenario is "through the fabric" transfer for isolated/private VMs where SSH/public IP connectivity isn't an option.

While we wait for a native Compute Service streaming API (which seems to be the long-term blocker), I propose we implement the **CLI-side orchestration** pattern that many users are heavily relying on manual scripts for.

### Proposed Implementation (Client-Side Bridge)

Instead of waiting for a new ARM API, the CLI can orchestrate the transfer using existing primitives:
1.  **Intermediate Storage:** CLI uses a storage account (specified by user or auto-discovered).
2.  **Upload:** CLI uploads the local file to a temp blob.
3.  **SAS Token:** CLI generates a short-lived (e.g., 10-15 min) SAS token.
4.  **Download on VM:** CLI executes `az vm run-command` to download the file inside the VM using the SAS URL.
5.  **Cleanup:** CLI deletes the temporary blob.

This pattern aligns with how other cloud CLIs (like `gcloud`) handle similar "magic" transfers—abstracting the complexity of intermediate storage from the user.

I am ready to start working on a PR for this "CLI-side orchestration" approach. Would the team be open to accepting this solution to finally unblock this request?

Thanks,
Ashutosh
